File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Sin.hs
module Glue.Lib.Math.Trigonometric.Sin where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Sine function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Sin.sin exactly
sin :: IR Eval
sin = NativeFunc sinImpl

-- Sine function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Sin.sinImpl exactly
sinImpl :: IR Eval -> Eval (IR Eval)
sinImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.sin (fromIntegral n))
    Float n -> pure $ Float (Prelude.sin n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/sin.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Sine function (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Sin.sin exactly
final Ir sin = IrNativeFunc(sinImpl);

/// Sine function implementation (radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Sin.sinImpl exactly
Eval<Ir> sinImpl(Ir arg) {
  return eval(arg).flatMap((va) {
    return switch (va) {
      IrInteger(value: final n) => Eval.pure(IrFloat(math.sin(n.toDouble()))),
      IrFloat(value: final n) => Eval.pure(IrFloat(math.sin(n))),
      _ => throwError(wrongArgumentType(['number'])),
    };
  });
}
